File: sdk/typescript/src/flow-commit.ts
import type { ConnectionLike } from './connection-types.js';
import type { AtomicFlowJobInput } from './flow-plan.js';
import type { JobRaw } from './job.js';
import type { DataResponse } from './responses.js';

interface PushFlowResult {
  jobs: JobRaw[];
}

export function validateFlowSnapshots(
  jobs: AtomicFlowJobInput[],
  snapshots: unknown
): Map<string, JobRaw> {
  if (!Array.isArray(snapshots) || snapshots.length !== jobs.length) {
    throw new Error('Invalid PUSHF response: committed job snapshots are missing');
  }
  const expected = new Map(jobs.map((job) => [job.id, job.queue]));
  const byId = new Map<string, JobRaw>();
  for (const value of snapshots) {
    if (!value || typeof value !== 'object') {
      throw new Error('Invalid PUSHF response: job snapshot is invalid');
    }
    const snapshot = value as JobRaw;
    if (typeof snapshot.id !== 'string' || !expected.has(snapshot.id)) {
      throw new Error('Invalid PUSHF response: committed job IDs do not match the request');
    }
    if (byId.has(snapshot.id) || snapshot.queue !== expected.get(snapshot.id)) {
      throw new Error('Invalid PUSHF response: committed job IDs do not match the request');
    }
    byId.set(snapshot.id, snapshot);
  }
  return byId;
}

export async function commitFlow(
  connection: ConnectionLike,
  jobs: AtomicFlowJobInput[]
): Promise<Map<string, JobRaw>> {
  if (jobs.length === 0) return new Map();
  const response = await connection.call<DataResponse<PushFlowResult>>({
    cmd: 'PUSHF',
    jobs,
  });
  return validateFlowSnapshots(jobs, response.data?.jobs);
}

File: sdk/rust/src/flow_commit.rs
use std::collections::HashMap;

use rmpv::Value;

use crate::flow::FlowNode;
use crate::flow_plan::{PlannedJob, PlannedNode};
use crate::wire::{Map, command, get, map};
use crate::{Connection, Error, Job, Result};

pub(crate) fn commit_flow(
    connection: &Connection,
    jobs: &[PlannedJob],
) -> Result<HashMap<String, Map>> {
    if jobs.is_empty() {
        return Ok(HashMap::new());
    }
    let request_jobs = jobs
        .iter()
        .map(|job| {
            Value::Map(map([
                ("id", Value::from(job.id.clone())),
                ("queue", Value::from(job.queue.clone())),
                ("input", Value::Map(job.input.clone())),
            ]))
        })
        .collect();
    let response = connection.call(command(
        "PUSHF",
        map([("jobs", Value::Array(request_jobs))]),
    ))?;
    let snapshots = get(&response, "data")
        .and_then(Value::as_map)
        .and_then(|data| get(data, "jobs"));
    validate_snapshots(jobs, snapshots)
}

pub(crate) fn validate_snapshots(
    jobs: &[PlannedJob],
    snapshots: Option<&Value>,
) -> Result<HashMap<String, Map>> {
    let snapshots = snapshots
        .and_then(Value::as_array)
        .filter(|items| items.len() == jobs.len())
        .ok_or_else(|| {
            Error::Serialization(
                "Invalid PUSHF response: committed job snapshots are missing".into(),
            )
        })?;
    let expected = jobs
        .iter()
        .map(|job| (job.id.as_str(), job.queue.as_str()))
        .collect::<HashMap<_, _>>();
    let mut by_id = HashMap::with_capacity(snapshots.len());
    for snapshot in snapshots {
        let raw = snapshot.as_map().ok_or_else(invalid_snapshot)?;
        let id = get(raw, "id")
            .and_then(Value::as_str)
            .ok_or_else(invalid_ids)?;
        let queue = get(raw, "queue")
            .and_then(Value::as_str)
            .ok_or_else(invalid_ids)?;
        if expected.get(id).copied() != Some(queue) || by_id.contains_key(id) {
            return Err(invalid_ids());
        }
        by_id.insert(id.to_owned(), raw.clone());
    }
    if by_id.len() != expected.len() {
        return Err(invalid_ids());
    }
    Ok(by_id)
}

pub(crate) fn build_node(
    node: &PlannedNode,
    snapshots: &HashMap<String, Map>,
    connection: &Connection,
) -> Result<FlowNode> {
    let raw = snapshots.get(&node.id).cloned().ok_or_else(invalid_ids)?;
    let children = node
        .children
        .iter()
        .map(|child| build_node(child, snapshots, connection))
        .collect::<Result<Vec<_>>>()?;
    Ok(FlowNode {
        job: Job::new(raw, connection.clone(), None),
        children,
    })
}

fn invalid_snapshot() -> Error {
    Error::Serialization("Invalid PUSHF response: job snapshot is invalid".into())
}

fn invalid_ids() -> Error {
    Error::Serialization(
        "Invalid PUSHF response: committed job IDs do not match the request".into(),
    )
}
